### PR TITLE
Add CSS variables - prerequisite for Dark mode 

### PR DESCRIPTION
**Pull request by timja**
See [JENKINS-XXXXX](https://issues.jenkins-ci.org/browse/JENKINS-XXXXX).

This is being used to deliver dark mode, see https://github.com/jenkinsci/dark-theme

## Testing

It would be great if someone can check this in IE 11, CSS variables aren't supported there but I've added a postcss plugin that has resolved the variables at build time, I've checked the properties are present in the CSS.

```
docker run --rm -ti -p 8080:8080 -e ID=4752 jenkins/core-pr-tester
```

### Proposed changelog entries

* Entry 1: Major RFE: Alpha dark theme released, see https://github.com/jenkinsci/dark-theme
* Entry 2: Developer: Add incubating support for CSS variables, useful in themes
* ...

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [x] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by res0nance** [comment]
I was going to test this on windows but with the current tip, it doesnt work on windows

```
[INFO] $ webpack --config webpack.config.js --mode=production
[ERROR] Browserslist: caniuse-lite is outdated. Please run next command `yarn upgrade`
[ERROR] webpack-fix-style-only-entries: removing js from style only module: base-styles-v2.js
[ERROR] webpack-fix-style-only-entries: removing js from style only module: ui-refresh-overrides.js
[INFO] Hash: 2b2ca561afb45798718d
[INFO] Version: webpack 4.41.2
[INFO] Time: 19069ms
[INFO] Built at: 05/27/2020 10:58:08 PM
[INFO]  51 assets
[INFO] Entrypoint page-init = vendors.js page-init.js
[INFO] Entrypoint pluginSetupWizard = vendors.js pluginSetupWizard.css pluginSetupWizard.js
[INFO] Entrypoint upgradeWizard = vendors.js upgradeWizard.js
[INFO] Entrypoint add-item = vendors.js add-item.css add-item.js
[INFO] Entrypoint config-scrollspy = vendors.js config-scrollspy.css config-scrollspy.js
[INFO] Entrypoint config-tabbar = vendors.js config-tabbar.css config-tabbar.js
[INFO] Entrypoint base-styles-v2 = vendors.js base-styles-v2.css
[INFO] Entrypoint ui-refresh-overrides = vendors.js
[INFO] [31] multi ./src/main/js/page-init.js 28 bytes {5} [built]
[INFO] [32] ./src/main/js/page-init.js 910 bytes {5} [built]
[INFO] [35] multi ./src/main/js/pluginSetupWizard.js ./src/main/less/pluginSetupWizard.less 40 bytes {6} [built]
[INFO] [41] ./src/main/less/pluginSetupWizard.less 711 bytes {6} [built]
[INFO] [43] multi ./src/main/js/upgradeWizard.js 28 bytes {8} [built]
[INFO] [44] ./src/main/js/upgradeWizard.js 1.4 KiB {8} [built]
[INFO] [45] multi ./src/main/js/add-item.js ./src/main/js/add-item.less 40 bytes {1} [built]
[INFO] [46] ./src/main/js/add-item.js 9.96 KiB {1} [built]
[INFO] [47] ./src/main/js/add-item.less 702 bytes {1} [built]
[INFO] [49] multi ./src/main/js/config-scrollspy.js ./src/main/js/config-scrollspy.less 40 bytes {3} [built]
[INFO] [50] ./src/main/js/config-scrollspy.js 4.54 KiB {3} [built]
[INFO] [51] ./src/main/js/config-scrollspy.less 710 bytes {3} [built]
[INFO] [53] multi ./src/main/js/config-tabbar.js ./src/main/js/config-tabbar.less 40 bytes {4} [built]
[INFO] [57] multi ./src/main/less/base-styles-v2.less 28 bytes {2} [built]
[INFO] [60] multi ./src/main/less/ui-refresh-overrides.less 28 bytes {7} [built]
[ERROR] error Command failed with exit code 2.
[INFO]     + 57 hidden modules
[INFO]
[INFO] ERROR in ./src/main/less/ui-refresh-overrides.less (./node_modules/mini-css-extract-plugin/dist/loader.js!./node_modules/css-loader/dist/cjs.js??ref--4-2!./node_modules/postcss-loader/src!./node_modules/less-loader/dist/cjs.js!./src/main/less/ui-refresh-overrides.less)
[INFO] Module build failed (from ./node_modules/mini-css-extract-plugin/dist/loader.js):
[INFO] ModuleNotFoundError: Module not found: Error: Can't resolve './abstracts/menu_down_arrow.png' in 'D:\github\jenkins\war\src\main\less'
[ERROR] error Command failed with exit code 2.
```

File: core/src/main/resources/lib/layout/breadcrumbs.css [hunk 1]
```diff
@@ -109,7 +109,7 @@
 
 #menuSelector {/* used for showing 'v' on the right of the anchor */
     background-color:transparent;
-    background-image: url(menu_down_arrow.png);
+    background-image: var(--menu-selector-down-background-image);
```
Comment (timja) [inline core/src/main/resources/lib/layout/breadcrumbs.css:112]: anyone able to check if this works on IE 11?

image


I suspect that it needs a manual fallback, looks like the postcss processor isn't adding the fallback for it

Comment (res0nance) [reply core/src/main/resources/lib/layout/breadcrumbs.css]: Seems to work on i.e but the entire dark theme doesn't

Here is a side by side with ie11 on the left and edge on the right

image

Comment (timja) [reply core/src/main/resources/lib/layout/breadcrumbs.css]: > Seems to work on i.e but the entire dark theme doesn't

It's not expected to work on IE 11 at all for the theme

> Here is a side by side with ie11 on the left and edge on the right

That's not working, at least for fonts, the preprocessor isn't picking them up 😢,

Two options I guess,

1. this work doesn't really require the fonts to be changeable at runtime, so it could go back to less variables
2. try out the other postcss plugin and see if that handles fonts, this one only seems to handle colours

cc @fqueiruga

Comment (res0nance) [reply core/src/main/resources/lib/layout/breadcrumbs.css]: Some more observations I made:
- The administrative monitor (The bell thing in the top right) still shows white background on the dark theme
- CLI page has black text with dark theme.

**Comment by fqueiruga** [comment]
I don't think we should go all in CSS variables. For example, I would keep font-family and font-size in less variables for now. 

In the first iteration I'd keep it simple: colors, backgrounds, borders and maybe font weights.

File: core/src/main/resources/lib/layout/breadcrumbs.css (hunk 1 above)
Comment (timja) [reply core/src/main/resources/lib/layout/breadcrumbs.css]: Thanks I’d forgotten about the admin monitor issue, and I hadn’t checked the CLI page

File: war/src/main/less/abstracts/font.less [hunk 2]
```diff
@@ -1,9 +1,12 @@
...
-@font-size-base: 1rem; // 16px
-@font-size-sm: 0.875rem; // 14px
-@font-size-xs: 0.75rem; // 12px
-@font-size-monospace: 0.95em;
-
-@line-height-base: 1.5;
-@line-height-heading: 1.2;
+body {
+  // Font related properties
+  --font-family-sans: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Noto Sans", Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
```
Comment (fqueiruga) [inline war/src/main/less/abstracts/font.less:3]: Maybe the fact that the variables are declared here and not in `:root` is causing the fallback issues. Just some brainstorming.

Comment (timja) [reply war/src/main/less/abstracts/font.less:3]: Yes you’re right

File: core/src/main/resources/lib/form/password/password.css [hunk 3]
```diff
@@ -27,9 +27,9 @@
 }
 
 .hidden-password-legend {
-    border: 1px solid #ccc;
+    border: 1px solid var(--input-border);
```
Comment (fqueiruga) [inline core/src/main/resources/lib/form/password/password.css:30]: For CSS adjuncts the fallback must be done by hand. For example, this case would be 

````
border: 1px solid #ccc;
border: 1px solid var(--input-border, #ccc);
````

This is one of the reasons of why I wanted to kill adjuncts.

File: core/src/main/resources/lib/layout/breadcrumbs.css [hunk 4]
```diff
@@ -19,7 +19,7 @@
     font-size: 1rem;
     line-height: 1.5;
     padding: 0 1.25rem;
-    background-color: #F8F8F8;
+    background-color: var(--breadcrumbs-bg);
```
Comment (fqueiruga) [inline core/src/main/resources/lib/layout/breadcrumbs.css:22]: This is an adjunct, fallback should be done by hand.

File: war/src/main/less/ui-refresh-overrides.less [hunk 5]
```diff
@@ -3,7 +3,7 @@
 
 .ui-refresh {
     .page-header {
-        background-color: @header-bg-v2;
+        background-color: var(--header-bg-v2);
```
Comment (fqueiruga) [inline war/src/main/less/ui-refresh-overrides.less:6]: I don't know if we should just override the `:root` variable here

File: war/src/main/less/pluginSetupWizard.less [hunk 6]
```diff
@@ -428,7 +428,7 @@
 						content: '';
 						left: 2px;
 						top: 3px;
-						background: @badge-color;
+						background: var(--badge-color);
```
Comment (fqueiruga) [inline war/src/main/less/pluginSetupWizard.less:431]: Is this variable declared anywhere? 

I wouldn't apply the styles to the setup wizard. Right know they are on their own and can behave weirdly if the variables are mixed.

File: war/src/main/less/modules/side-panel-tasks.less [hunk 7]
```diff
@@ -44,12 +44,12 @@
...
 .task-link:link,
 .task-link:visited {
-  color: @text-color;
+  color: var(--text-color);
   text-decoration: none;
 
   &:hover,
   &:focus {
-    background-color: #F8F8F8;
+    background-color: var(--task-link-background-color--hover);
```
Comment (fqueiruga) [inline war/src/main/less/modules/side-panel-tasks.less:52]: I'm thinking that maybe we should just treat these links and the ones in the footer as a dark link variant. They all behave the same, and I don't think we should be hyper specific here.

File: war/src/main/less/abstracts/font.less (hunk 2 above)
Comment (fqueiruga) [reply war/src/main/less/abstracts/font.less:3]: Maybe we should just go ahead and create a big & comprehensive `variables.less` file and be done with it :D

File: war/src/main/less/abstracts/colors.less [hunk 8]
```diff
@@ -1,84 +1,150 @@
...
+  --project-status-border-bottom-color: #ddd;
+  --table-parameters-background-color--hover: #f0f0f0;
+  --table-stripped-background-color--hover: #e8e8e8;
+
+  // Link
+  --link-color: #204a87;
+  --link-visited-color: #5c3566;
+
+  // Pane
+  --pane-header-color: #eee;
```
Comment (fqueiruga) [inline war/src/main/less/abstracts/colors.less:111]: Does this apply to all panes? For example, in a future PR I was going to change it just for side-panel widgets, dunno if this would affect it.

File: war/postcss.config.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+  parser: "postcss-less",
+  plugins: [
+    require('autoprefixer'),
+    require('postcss-custom-properties')
```
Comment (fqueiruga) [inline war/postcss.config.js:5]: Does the plugin order matter? I'm thinking if there can be a scenario where the CSS variables fallback is applied for a main CSS rule but not for the vendor prefixes.

File: war/src/main/less/base/style.less [hunk 10]
```diff
@@ -1461,7 +1455,32 @@ DIV.yahooTree td {
 
 #jenkins .yuimenu .bd {
     border: 1px solid #ccc;
-    box-shadow: 0px 3px 10px #bbb;
+    box-shadow: var(--menu-box-shadow);
+}
+
+#jenkins.yui-skin-sam {
```
Comment (fqueiruga) [inline war/src/main/less/base/style.less:1461]: Why did you need to create this? Does this fix something?

File: core/src/main/resources/lib/layout/breadcrumbs.css (hunk 1 above)
Comment (fqueiruga) [reply core/src/main/resources/lib/layout/breadcrumbs.css]: BTW I would not bother converting the arrow to a CSS variables. It should be either an icon or a CSS triangle, not an image anyway.

**Comment by oleg-nenashev** [comment]
Deployed the current commit to https://repo.jenkins-ci.org/snapshots/org/jenkins-ci/main/jenkins-war/2.239-SNAPSHOT/jenkins-war-2.239-20200527.181313-1.war , for testing in https://github.com/jenkinsci/dark-theme/pull/8

File: war/src/main/less/base/style.less (hunk 10 above)
Comment (timja) [reply war/src/main/less/base/style.less:1461]: it allows overriding the yui-skin in themes, I've matched the css below to what the yui skin provides and extended it with variables

File: war/postcss.config.js (hunk 9 above)
Comment (timja) [reply war/postcss.config.js:5]: isn't the postcss-custom-properties applied second so it'll be applied on top of the prefixed?

File: war/src/main/less/abstracts/colors.less (hunk 8 above)
Comment (timja) [reply war/src/main/less/abstracts/colors.less:111]: it's applied to `.pane-header, .pane-footer`

File: war/src/main/less/ui-refresh-overrides.less (hunk 5 above)
Comment (timja) [reply war/src/main/less/ui-refresh-overrides.less:6]: hmm? not quite sure what you mean

File: war/src/main/less/modules/side-panel-tasks.less (hunk 7 above)
Comment (timja) [reply war/src/main/less/modules/side-panel-tasks.less:52]: they are used for different parts of the interface though, one is used for background colour and this is used for hover, bit wary of overloading variables, internally they are re-using the same colour variable, but they are different parts of the interface...

Comment (fqueiruga) [reply war/src/main/less/modules/side-panel-tasks.less:52]: I didn't see this was about bg color nevermind. My comments were about color and font-weight basically.

File: war/postcss.config.js (hunk 9 above)
Comment (fqueiruga) [reply war/postcss.config.js:5]: No idea, that's why I was asking.

Comment (timja) [reply war/postcss.config.js:5]: I think this is okay

**Comment by fqueiruga** [comment]
There are some variables that use `-background-` and some that use `-bg-`. I think we should standardize on one. I'm partial towards `-bg-` but I'm not really invested on it.

File: core/src/main/resources/lib/layout/breadcrumbs.css (hunk 1 above)
Comment (timja) [reply core/src/main/resources/lib/layout/breadcrumbs.css]: https://issues.jenkins-ci.org/browse/JENKINS-62496

**Comment by timja** [comment]
> There are some variables that use `-background-` and some that use `-bg-`. I think we should standardize on one. I'm partial towards `-bg-` but I'm not really invested on it.

Agreed, most of the bg ones came from the existing sass variables and I didn't touch them in the find / replace,

When I created new ones I added them with background, happy to shorten though

**Comment by fqueiruga** [comment]
Created a PR against this branch https://github.com/timja/jenkins/pull/3

**Comment by fqueiruga** [comment]
Another PR created against this branch https://github.com/timja/jenkins/pull/4

**Comment by fqueiruga** [comment]
Created another PR, forgot one variable https://github.com/timja/jenkins/pull/6 😓

**Comment by fqueiruga** [comment]
I'll start working on the dark mode changes for these variables

**Comment by res0nance** [comment]
Confirming that IE11 works fine with https://github.com/jenkinsci/jenkins/pull/4752/commits/670b90776b8373576ac51482c2080cba2dba4e1c

File: war/src/main/less/abstracts/colors.less [hunk 11]
```diff
@@ -1,84 +1,151 @@
...
+  --tab-bar-border-color--checked: #f0f0f0;
+  --tab-bar-text-color--checked: #000;
+  --tab-bar-bg-color--hover: #eee;
+  --tab-bar-add-text-color: #0460d1;
+  --tab-bar-add-bg-color--hover: #d8efff;
+  --tab-bar-add-border-color: #d8efff;
+
+  // Side panel
+  --panel-border-color: var(#eaeff2);
+  --panel-border-color--hover: var(#cecece);
```
Comment (fqueiruga) [inline war/src/main/less/abstracts/colors.less:132]: These rules were not having effect
```suggestion
  --panel-border-color: #eaeff2;
  --panel-border-color--hover: #cecece;
```

**Comment by fqueiruga** [comment]
Can we add the squash-merge-me label?

**Review by fqueiruga — CHANGES_REQUESTED** [review]

**Review by fqueiruga — APPROVED** [review]

**Review by EstherAF — CHANGES_REQUESTED** [review]

File: core/src/main/resources/lib/layout/breadcrumbs.css [hunk 12]
```diff
@@ -66,6 +67,7 @@
     display: inline-block;
     display: inline-flex;
     color: #4d545d;
+    color: var(--breadcrumbs-color--visited);
```
[the remark this answers is not in the document]
Comment (timja) [reply core/src/main/resources/lib/layout/breadcrumbs.css:70]: thanks, checked and you're right!

**Comment by fqueiruga** [comment]
Shouldn't this have a RFE label?

**Review by oleg-nenashev — APPROVED** [review]
Thanks to all contributors! The Dark theme looks pretty good with the recent improvements, and it should greatly simplify development and maintenance of themes in the future. I am +100 for merging, and I suggest converting it to a major RFE and hinting the Dark Theme prototype there. It would be great to facilitate the feedback and adoption in this area after the changes.

P.S: I have also tested the current patch with https://tobix.github.io/jenkins-neo2-theme/ , and it looks pretty well after the changes. FYI @TobiX

**Comment by timja** [comment]
I've updated the changelog PTAL

**Review by res0nance — APPROVED** [review]
Retested up to https://github.com/jenkinsci/jenkins/pull/4752/commits/c12b42c95a019dbc0074c7ad8f0181a116eb0a6e
With IE11 things look ok there

**Comment by timja** [comment]
Some shuffling done in https://github.com/jenkinsci/jenkins/pull/4752/commits/49e501f37db848e69de31607a43df79bab4e77b0 mostly moves yui customisation code to one place, as there was some in styles.less as well

I removed a bit of duplication between the search bar and other autocompletes, and pulled in some of the enhancements that were added to search but not others.

I've tested it on the add job, copy items from widget

**Comment by oleg-nenashev** [comment]
We may merge it in 24 hours if no negative feedback or requests to extend the review time. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information

Thanks to @timja @95jonpet @fqueiruga @res0nance @EstherAF @nimishbongale who contributed to the [Dark Theme](https://github.com/jenkinsci/dark-theme) implementation and evaluation, and to compatibility testing of this pull request! It is great to see such a major improvement to Jenkins UX and CSS extensibility delivered in a few days. 🥇 🙇 🚀

**Comment by oleg-nenashev** [comment]
https://github.com/jenkinsci/dark-theme/pull/60 updates the demo for testing to the recent version

File: war/postcss.config.js [hunk 13]
```diff
@@ -1,6 +1,7 @@
 module.exports = {
   parser: "postcss-less",
   plugins: [
-    require('autoprefixer')
+    require('autoprefixer'),
+    require('postcss-custom-properties')
```
Comment (halkeye) [inline war/postcss.config.js:5]: so reading up on what this does. Doesn't custom-properties plugin remove all the variables and replace them with the actual values?

File: war/src/main/js/widgets/add/addform.less [hunk 14]
```diff
@@ -4,9 +4,9 @@
 
   form > .header {
     position: relative;
-    border: 1px solid #cccccc;
```
Comment (halkeye) [inline war/src/main/js/widgets/add/addform.less:7]: shouldn't this fallback be kept?

**Comment by halkeye** [comment]
so not a blocker but a few concerns:

1) Sometimes the fallback values before the variables are kept like
https://github.com/jenkinsci/jenkins/pull/4752/files#diff-21e1a8d53a5c4282d7a7c221551cf7f9L22-R24
But other times they are not
Since the postcss plugin should replace the variables at compile time, i don't think the fallbacks are necessary and will get out of sync/confusion quickly
2) I haven't checked but does the plugin keep the variables in the final output? I would assume so because they would make themeing so much easier. If not, can it be configured to do the replacements and keep them?

File: war/postcss.config.js (hunk 13 above)
Comment (timja) [reply war/postcss.config.js:5]: No it doesn’t remove them it adds the fallbacks above the variables for browsers that don’t support them (IE)

The theme wouldn’t work at all if it did it in the way you described

File: war/src/main/js/widgets/add/addform.less (hunk 14 above)
Comment (timja) [reply war/src/main/js/widgets/add/addform.less:7]: Post css handles that

**Comment by timja** [comment]
1. The manual fallbacks are kept in the adjuncts which aren’t passed through webpack / postcss
2. Replied to this in a different comment but the fallbacks are added above the variables in the final output, and are only used if a browser doesn’t support variables (Internet explorer)

File: war/src/main/less/abstracts/colors.less [hunk 15]
```diff
@@ -1,84 +1,181 @@
...
+  --tab-bar-text-color--checked: #000;
+  --tab-bar-bg-color--hover: #eee;
+  --tab-bar-add-text-color: #0460d1;
+  --tab-bar-add-bg-color--hover: #d8efff;
+  --tab-bar-add-border-color: #d8efff;
+
+  // Side panel
+  --panel-border-color: #eaeff2;
+  --panel-border-color--hover: #cecece;
+  --side-panel-hover-color: var(--panel-border-color);
```
Comment (daniel-beck) [inline war/src/main/less/abstracts/colors.less:142]: Defined but never used, also naming inconsistent with other variables.

File: war/src/main/less/modules/buttons.less [hunk 16]
```diff
@@ -1,77 +1,77 @@
...
   }
 }
 
 .button-primary() {
-  color: @white;
-  background-color: @btn-primary-bg;
-  border-color: @btn-primary-bg;
+  color: var(--btn-text-color);
+  background-color: var(--btn-primary-bg);
+  border-color: var(--btn-primary-bg);
```
Comment (daniel-beck) [inline war/src/main/less/modules/buttons.less:28]: Is it deliberate that here and further below, we're not using a separate variable for the primary button border? Secondary buttons have a separate variable.

Comment (timja) [reply war/src/main/less/modules/buttons.less:28]: historical most likely, not intentional as far as I know

File: war/src/main/less/abstracts/colors.less (hunk 15 above)
Comment (fqueiruga) [reply war/src/main/less/abstracts/colors.less:142]: Should be removed IMO

File: war/src/main/less/modules/buttons.less (hunk 16 above)
Comment (fqueiruga) [reply war/src/main/less/modules/buttons.less:28]: Button variable API should be expanded IMO
